src/persist-client/src/examples/persistent_source_example.rs

```diff
@@ -173,6 +182,33 @@ mod api {
             }
         }
     }
+
+    // TODO: Obviously a placeholder, in the real world, there would be timeouts, and maybe a
+    // `Result` return type.
+    //
+    // Also, the updates are not differential in this interface, while they probably should be.
+    // As it is now, we can never retract updates.
+    #[async_trait]
+    pub trait IteratedConsensus<K, V, T>: Clone
+    where
+        T: Clone,
+    {
+        /// Returns all updates with timestamps up to (and including) the given `as_of`.
+        ///
+        /// Returns an `Err` with the current upper if the given `as_of` is beyond the current
+        /// upper.
+        async fn snapshot(&self, as_of: Antichain<T>) -> Result<Vec<((K, V), T)>, Antichain<T>>;
+
+        /// Tries to append the given updates to the maintained collection. If the current
+        /// upper is not `expected_upper` this will return an `Err` containing the current
+        /// upper.
+        async fn append(
```

Comment:
sorry to nit but please let's not call this `append`, which is already a thing :). I'm fine with something like `compare_and_set`